Unify socket into file system to avoid the code duplication

Donny9 opened the pull request:
## Summary
Unify socket into file system to avoid the code duplication. #2833 this issue link associated with this patch.
Key modification points:
1. This patch refines the interaction between fs and net, we will address socket api by fs api, example:
fs:       ->          socket:
socket->...->psock_socket
close  ->...->psock_close
write   ->...->psock_send
read   ->...->psock_recv
ioctl    ->...->psock_ioctl
poll     ->...->psock_poll
dup    ->...->psock_dup2

2. this patch  deletes the wraper api net_xxx, eg:net_close，net_dup，and so on to avoid the code duplication.

3. This patch delete config CONFIG_NSOCKET_DESCRIPTORS and group.tg_socketlist. 

4. Bind the socket to each file and set file->f_priv to socket.
## Impact
Reduce duplication code and unified management of socket and fs.
## Testing
daily test.

gustavonihei commented on include/nuttx/net/net.h:
```diff
@@ -1333,87 +1240,27 @@ struct pollfd; /* Forward reference -- see poll.h */
  *
  * Description:
- *   Clone a socket descriptor to an arbitrary descriptor number.
+ *   Clone a socket instance to an new instance.
```
```suggestion
 *   Clone a socket instance to a new instance.
```

Donny9 replied:
Ok, thanks~

gustavonihei commented on include/nuttx/net/net.h:
```diff
@@ -1333,87 +1240,27 @@ struct pollfd; /* Forward reference -- see poll.h */
  * Input Parameters:
- *   sockfd - Socket descriptor of the socket to operate on
+ *   psock  - An instance of the internal socket structure.
  *   buf    - Caller-provided location in which to return the fstat data
```
```suggestion
 *   buf    - Caller-provided location in which to return the fstat data.
```
nit: Add a period for consistency

Donny9 replied:
Ok, thanks

gustavonihei commented on include/sys/poll.h:
```diff
@@ -78,8 +78,7 @@
 #define POLLFILE     (0x40)
-#define POLLSOCK     (0x80)
-#define POLLMASK     (0xC0)
+#define POLLMASK     (0x40)
```
I think it is better if `POLLMASK` is defined as an OR operation of all previously defined event types.
```suggestion
#define POLLMASK     (POLLFD | POLLFILE)
```

Donny9 replied:
keep for consistency?

gustavonihei replied:
Sorry, I don't understand how the proposed approach is less consistent than the current one.
In fact, I think it is kinda normal to define a mask based on the existing values the it should "mask".

Donny9 replied:
okay, done!

gustavonihei commented on net/pkt/pkt_send.c:
```diff
@@ -162,7 +162,7 @@ ssize_t psock_pkt_send(FAR struct socket *psock, FAR const void *buf,
   /* Verify that the sockfd corresponds to valid, allocated socket */
 
-  if (!psock || psock->s_crefs <= 0)
+  if (!psock || !psock->s_conn)
```
Throughout the PR, `psock->s_conn` is both being check against NULL (which I prefer), and used as input to operator `!`.
As a suggestion, it is better to always compare it with NULL, since `s_conn` is in fact a pointer.

Donny9 replied:
Done!

gustavonihei commented on net/socket/accept.c:
```diff
@@ -284,7 +278,7 @@ int accept(int sockfd, FAR struct sockaddr *addr, FAR socklen_t *addrlen)
 
 errout_with_socket:
-  sockfd_release(newfd);
+  close(newfd);
```
We should not call the close(), since it sets a cancellation point.
```suggestion
  nx_close(newfd);
```

Donny9 replied:
Done.

gustavonihei commented on fs/socket/socket.c:
```diff
@@ -0,0 +1,298 @@
+  return sockfd;
+
+errout_with_sockfd:
+  close(sockfd);
```
We should not call the close(), since it sets a cancellation point.
```suggestion
  nx_close(sockfd);
```

Donny9 replied:
Done!

patacongo commented:
Should not this also be updated/removed:  https://github.com/patacongo/incubator-nuttx/blob/master/TODO#L1902

Donny9 commented:
> Should not this also be updated/removed: https://github.com/patacongo/incubator-nuttx/blob/master/TODO#L1902
> 
> Perhaps a better thing to do would be to convert all items in the TODO list to either issues or to Wiki pages???
> 
> There is also some crazy using pipes to account for the the fact that the stdin, stdout, stderr cannot be sockets in apps/netutils/thttpd. Should that also not be corrected?



> Should not this also be updated/removed: https://github.com/patacongo/incubator-nuttx/blob/master/TODO#L1902
> 
> Perhaps a better thing to do would be to convert all items in the TODO list to either issues or to Wiki pages???
> 
> There is also some crazy using pipes to account for the the fact that the stdin, stdout, stderr cannot be sockets in apps/netutils/thttpd. Should that also not be corrected?

Okay, i have modified it, and modified nuttx/app in patch:https://github.com/apache/incubator-nuttx-apps/pull/601. Thanks~

Donny9 commented:
> Should not this also be updated/removed: https://github.com/patacongo/incubator-nuttx/blob/master/TODO#L1902
> 
> Perhaps a better thing to do would be to convert all items in the TODO list to either issues or to Wiki pages???
> 
> There is also some crazy using pipes to account for the the fact that the stdin, stdout, stderr cannot be sockets in apps/netutils/thttpd. Should that also not be corrected?

Okay, i have modified it, and modified nuttx/app in patch:apache/incubator-nuttx-apps#601. Thanks~

davids5 commented:
@Donny9 - Nice - how much smaller is this then master?

Donny9 commented:
> @Donny9 - Nice - how much smaller is this then master?

I use config stm32f4discovery:ipv6 to compare this change, and use bloaty tool to analyse.
nuttx$ bloaty -n 20 -d sections,symbols this_patch -- master                                                                                                                                
    FILE SIZE        VM SIZE                                                                                                                                                                                
 --------------  --------------                                                                                                                                                                             
  +3.4%     +28  +3.6%     +28    .data                                                                                                                                                                     
    [NEW]     +24  [NEW]     +24    g_sock_inode                                                                                                                                                            
    +100%      +2  +100%      +2    g_devno_sem                                                                                                                                                             
    +100%      +2  +100%      +2    g_spawn_parmsem                                                                                                                                                         
  -1.4%      -2  [ = ]       0    .shstrtab                                                                                                                                                                 
  -2.6%     -24  -2.7%     -24    .ARM.extab                                                                                                                                                                
  -1.0%     -80  -1.0%     -80    .ARM.exidx                                                                                                                                                                
  -0.5%    -154  [ = ]       0    .strtab                                                                                                                                                                   
    [NEW]     +21  [ = ]       0    g_last_tcp_port.5480                                                                                                                                                    
    [NEW]     +21  [ = ]       0    g_last_udp_port.5260                                                                                                                                                    
    [NEW]     +18  [ = ]       0    __FUNCTION__.5936                                                                                                                                                       
    [NEW]     +18  [ = ]       0    __FUNCTION__.6321                                                                                                                                                       
    [NEW]     +18  [ = ]       0    __FUNCTION__.6329                                                                                                                                                       
    [NEW]     +18  [ = ]       0    __FUNCTION__.6337                                                                                                                                                       
    [NEW]     +18  [ = ]       0    __FUNCTION__.6347                                                                                                                                                       
    [NEW]     +18  [ = ]       0    notifier_key.6174                                                                                                                                                       
   -43.2%     -16  [ = ]       0    g_icmpv6_connections                                                                                                                                                    
    [DEL]     -16  [ = ]       0    net_releaselist                                                                                                                                                         
    [DEL]     -18  [ = ]       0    __FUNCTION__.5971                                                                                                                                                       
    [DEL]     -18  [ = ]       0    __FUNCTION__.6356                                                                                                                                                       
    [DEL]     -18  [ = ]       0    __FUNCTION__.6364                                                                                                                                                       
    [DEL]     -18  [ = ]       0    __FUNCTION__.6372                                                                                                                                                       
    [DEL]     -18  [ = ]       0    __FUNCTION__.6382                                                                                                                                                       
    [DEL]     -18  [ = ]       0    notifier_key.6209                                                                                                                                                       
    [DEL]     -20  [ = ]       0    nxsched_get_sockets                                                                                                                                                     
    [DEL]     -21  [ = ]       0    g_last_tcp_port.5512
    [DEL]     -21  [ = ]       0    g_last_udp_port.5292
    -7.7%     -33  [ = ]       0    [38 Others]
    -0.8%     -69  [ = ]       0    [section .strtab]
  -0.3%    -224  [ = ]       0    .symtab
    [NEW]     +48  [ = ]       0    __FUNCTION__.5936
    [NEW]     +48  [ = ]       0    buffer.5951
    [NEW]     +48  [ = ]       0    g_sock_inode
    [NEW]     +48  [ = ]       0    inited.5901
    [NEW]     +32  [ = ]       0    __FUNCTION__.6321
    [NEW]     +32  [ = ]       0    g_last_udp_port.5260
    [NEW]     +32  [ = ]       0    g_sock_fileops
     +40%     +32  [ = ]       0    icmpv6_input
    [NEW]     +32  [ = ]       0    labels.6531
    [DEL]     -32  [ = ]       0    __FUNCTION__.6356
    [DEL]     -32  [ = ]       0    g_last_udp_port.5292
    [DEL]     -32  [ = ]       0    labels.6566
    [DEL]     -32  [ = ]       0    net_checksd
    [DEL]     -32  [ = ]       0    net_close
    [DEL]     -32  [ = ]       0    net_dup
    [DEL]     -48  [ = ]       0    __FUNCTION__.5971
    [DEL]     -48  [ = ]       0    buffer.5986
    [DEL]     -48  [ = ]       0    inited.5936
    [DEL]     -48  [ = ]       0    nxsched_get_sockets
    -7.8%     -64  [ = ]       0    [47 Others]
    -0.8%    -128  [ = ]       0    [section .symtab]
  -0.4%    -504  -0.4%    -504    .text
    [NEW]     +46  [NEW]     +46    sock_file_open
    [NEW]     +44  [NEW]     +44    g_sock_fileops
     +56%     +28   +56%     +28    nx_vfcntl
    +1.9%     +24  +1.9%     +24    proc_read
    [NEW]     +20  [NEW]     +20    sock_file_close
    [NEW]     +20  [NEW]     +20    sock_file_read
    [DEL]     -18  [DEL]     -18    net_dup
    [DEL]     -22  [DEL]     -22    net_vfcntl
    [DEL]     -22  [DEL]     -22    netdev_vioctl
    [DEL]     -22  [DEL]     -22    poll_fdsetup
    [DEL]     -26  [DEL]     -26    psock_release
    [DEL]     -30  [DEL]     -30    net_checksd
    [DEL]     -34  [DEL]     -34    net_releaselist
    [DEL]     -38  [DEL]     -38    net_poll
    [DEL]     -40  [DEL]     -40    sockfd_release
    -0.6%     -42  -0.6%     -42    [section .text]
   -25.0%     -46 -25.0%     -46    psock_vfcntl
   -14.8%     -62 -14.8%     -62    nx_poll
    [DEL]     -76  [DEL]     -76    net_dup2
    [DEL]     -90  [DEL]     -90    psock_dup
    -3.3%    -118  -3.3%    -118    [60 Others]
  -0.4%    -960  -0.4%    -580    TOTAL

And this patch run in sim, memory usage:

this patch:
nsh> free
             total       used       free    largest
Umem:     67108864     378080   66730784   66730304

master:
nsh> free
             total       used       free    largest
Umem:     67108864     381568   66727296   66726816

reduce (381568-378080)/1024=3.4kbytes

xiaoxiang781216 commented:
This patch pass our internal test, does anyone have more concern to address?

masayuki2009 commented:
@Donny9 @xiaoxiang781216 

Hmmm, it seems that the telnetd does not work after this change.
You can try lm3s6965-ek:discover with qemu.

```
nsh> renew eth0
nsh> ps
  PID PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK   STACK   USED  FILLED COMMAND
    0   0 FIFO     Kthread N-- Ready              00000000 001024 001024 100.0%! Idle Task
    1 224 RR       Kthread --- Waiting  Signal    00000000 002036 000236  11.5%  hpwork
    2  60 RR       Kthread --- Waiting  Signal    00000000 002036 000380  18.6%  lpwork
    3 100 RR       Task    --- Running            00000000 002036 001000  49.1%  init
    4 100 RR       Task    --- Waiting  Semaphore 00000000 002012 000420  20.8%  Telnet daemon 0x20006050
nsh> ifconfig
eth0	Link encap:Ethernet HWaddr 00:e0:de:ad:be:ef at UP
	inet addr:10.0.2.15 DRaddr:10.0.2.2 Mask:255.255.255.0

             IPv4   TCP   UDP  ICMP
Received     0002  0000  0002  0000
Dropped      0000  0000  0000  0000
  IPv4        VHL: 0000   Frg: 0000
  Checksum   0000  0000  0000  ----
  TCP         ACK: 0000   SYN: 0000
              RST: 0000  0000
  Type       0000  ----  ----  0000
Sent         0002  0000  0002  0000
  Rexmit     ----  0000  ----  ----
nsh> qemu: fatal: Lockup: can't escalate 3 to HardFault (current priority -1)

R00=00000000 R01=00000000 R02=40025000 R03=00000004
R04=00000000 R05=00000003 R06=ffffff6c R07=2000724c
R08=20006178 R09=00000064 R10=00000400 R11=00000001
R12=00000074 R13=ffffff50 R14=00000c7f R15=00000000
XPSR=00000003 ---- A handler
FPSCR: 00000000
Aborted (core dumped)

$ telnet localhost 10023
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
Connection closed by foreign host.

```

masayuki2009 commented:
@xiaoxiang781216 @Donny9 

I think we should revert this PR until the above issue is fixed.

xiaoxiang781216 commented:
We are looking the issue, please give us sometime to investigate. let's revert it if we can't fix it today.

Donny9 commented:
> lm3s6965-ek:discover

how to run lm3s6965-ek:discover on qemu? use command: qemu-system-arm -semihosting -M lm3s6965evb -device loader,file=nuttx.bin,addr=0x00000000 -netdev user,id=user0 -nographic -kernel nuttx ? I don't connect it with arm-none-eabi-gdb. please tell me the way. Thank you @masayuki2009

masayuki2009 commented:
> > lm3s6965-ek:discover
> 
> how to run lm3s6965-ek:discover on qemu? use command: qemu-system-arm -semihosting -M lm3s6965evb -device loader,file=nuttx.bin,addr=0x00000000 -netdev user,id=user0 -nographic -kernel nuttx ? I don't connect it with arm-none-eabi-gdb. please tell me the way. Thank you @masayuki2009

@Donny9 

$ qemu-system-arm -net nic,model=stellaris -net user,hostfwd=tcp:127.0.0.1:10023-10.0.2.15:23,hostfwd=tcp:127.0.0.1:10021-10.0.2.15:21 -M lm3s6965evb -kernel nuttx -nographic -s -S

$ arm-none-eabi-gdb-py -i=mi -ix=./tools/nuttx-gdbinit nuttx
(gdb) target extended-remote:1234

Donny9 commented:
> > > lm3s6965-ek:discover
> > 
> > 
> > how to run lm3s6965-ek:discover on qemu? use command: qemu-system-arm -semihosting -M lm3s6965evb -device loader,file=nuttx.bin,addr=0x00000000 -netdev user,id=user0 -nographic -kernel nuttx ? I don't connect it with arm-none-eabi-gdb. please tell me the way. Thank you @masayuki2009
> 
> @Donny9
> 
> Firstly, you need to add `CONFIG_TIVA_WITH_QEMU=y`
> 
> $ qemu-system-arm -net nic,model=stellaris -net user,hostfwd=tcp:127.0.0.1:10023-10.0.2.15:23,hostfwd=tcp:127.0.0.1:10021-10.0.2.15:21 -M lm3s6965evb -kernel nuttx -nographic -s -S
> 
> $ arm-none-eabi-gdb-py -i=mi -ix=./tools/nuttx-gdbinit nuttx
> (gdb) target extended-remote:1234
> 
> $ telnet localhost 10023

hello, @masayuki2009, please try this patch #2978 thanks.
